### DISCUSSION TITLE
Why `stylex.create` needs to be defined at the top level

### DISCUSSION THREAD

**Discussion by nonzzz** [facebook/stylex#154] [Q&A] [answered]
I try using stylex in vue setup. but i got a parsaer error is that `stylex.create() is only allowed at the root of a program.`

Here is transformed code by vue:

```js

import _export_sfc from 'plugin-vue:export-helper'

const _sfc_main = {
  __name: 'app',
  setup(__props, { expose: __expose }) {
    __expose()

    const s = stylex.create({
      blue: {
        color: 'blue'
      }
    })

    const cl = computed(() => stylex.props(styles.normal).className)

    const __returned__ = { s, cl, get stylex() { return stylex }, computed, get styles() { return styles }, Card }
    Object.defineProperty(__returned__, '__isScriptSetup', { enumerable: false, value: true })
    return __returned__
  }

}

function _sfc_render(_ctx, _cache, $props, $setup, $data, $options) {
  return (_openBlock(), _createElementBlock('div', {
    class: _normalizeClass($setup.cl)
  }, [
    _createVNode($setup.Card, null, {
      default: _withCtx(() => [
        _createTextVNode(' 123 456 ')
      ]),
      _: 1 /* STABLE */
    })
  ], 2 /* CLASS */))
}

...

```

I checked the source code and noticed,but i don't understand why verificaton is needed.

https://github.com/facebook/stylex/blob/1f00fe6fd943f658d2282dccf2aa337f4eb7b339/packages/babel-plugin/src/visitors/stylex-create/index.js#L234-L239

**Comment by nmn** [comment, marked as answer]
The `stylex.props` function uses memoization at runtime. This requires the objects generated by `stylex.create` to maintain the same reference.

Defining styles *within* a function body would mean that the style objects are re-created on every render, or at least for each instance of the component. This is unnecessarily inefficient.

Further, being in the scope of a function, you might intuitively try to use variables and other dynamic values within `stylex.create` which is not supported because the compiler needs to know the styles created at compile time.

---

In any case, in Vue, instead of using `stylex.create` in a `<script setup>`, use it in a plain `<script>` instead and it'll be at the top level.

I imagine, you tried something like this:

```html
<script setup>
import { ref } from 'vue'

import * as stylex from '@stylexjs/stylex'

const s = stylex.create({
  blue: {
    color: 'blue'
  }
})

const msg = ref('Hello World!');
const count = ref(0);
</script>
<template> ... </template>
```

Instead do this:

```html
<script>
import * as stylex from '@stylexjs/stylex'

const s = stylex.create({
  blue: {
    color: 'blue'
  }
})
</script>

<script setup>
import { ref } from 'vue'

const msg = ref('Hello World!');
const count = ref(0);
</script>

<template> ... </template>
```

I.e. The StyleX styles will be defined in a block separately from your state. The same is true with Svelte. You can also put this block at the bottom of the file, after the `<template>`.

```html
<script setup>
// State
import { ref } from 'vue'

const msg = ref('Hello World!');
const count = ref(0);
</script>

<template> ... </template>

<script>
// Styles:
import * as stylex from '@stylexjs/stylex'

const s = stylex.create({
  blue: {
    color: 'blue'
  }
})
</script>
```

**Reply by nonzzz** [reply]
Thanks!
